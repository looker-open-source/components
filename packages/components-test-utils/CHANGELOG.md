# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.10.4](https://github.com/looker-open-source/components/compare/v0.10.3...v0.10.4) (2021-01-25)

**Note:** Version bump only for package @looker/components-test-utils





## [0.10.3](https://github.com/looker-open-source/components/compare/v0.10.2...v0.10.3) (2021-01-25)

**Note:** Version bump only for package @looker/components-test-utils





## [0.10.2](https://github.com/looker-open-source/components/compare/v0.10.1...v0.10.2) (2021-01-22)


### Bug Fixes

* On build properly set package.json "files" key ([#1806](https://github.com/looker-open-source/components/issues/1806)) ([c81c4bb](https://github.com/looker-open-source/components/commit/c81c4bb625c58ede49957aad90ac8d9d7b2c4b79))





## [0.10.1](https://github.com/looker-open-source/components/compare/v0.10.0...v0.10.1) (2021-01-22)


### Bug Fixes

* Properly build ems/cjs/ts for easy deep-imports ([#1804](https://github.com/looker-open-source/components/issues/1804)) ([2a523cd](https://github.com/looker-open-source/components/commit/2a523cd70b079944376c6dc87c18202e05a97b01))





# [0.10.0](https://github.com/looker-open-source/components/compare/v0.9.29...v0.10.0) (2021-01-19)


### Bug Fixes

* Properly configure build for tree-shaking support ([9202b72](https://github.com/looker-open-source/components/commit/9202b72bc5fb99eb3e1af7d9e56f3dc15b1df2a4))





## [0.9.15]

### Changed

- `@looker/components-test-utils` helpers will no longer produce `globalStyles`
