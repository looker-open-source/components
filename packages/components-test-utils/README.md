# Test Utilities

Shared functions used for mocking and testing Looker UI Components.
