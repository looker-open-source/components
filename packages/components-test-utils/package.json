{
  "name": "@looker/components-test-utils",
  "license": "MIT",
  "version": "0.9.12",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/looker-open-source/components",
    "directory": "packages/components-test-utils"
  },
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "@looker/components-providers": "^0.9.12",
    "@testing-library/react": "^10.4.8",
    "@types/enzyme": "^3.10.5",
    "@types/jest": "^26.0.10",
    "@types/react": "^16.9.49",
    "@types/react-test-renderer": "^16.9.3",
    "enzyme": "^3.11.0",
    "jest-styled-components": "^6.3.4",
    "react": "^16.13.1",
    "react-test-renderer": "^16.13.1",
    "styled-components": "^4.4.1"
  },
  "peerDependencies": {
    "react": "^16.11"
  },
  "gitHead": "62febc162e3e45124f403e2c84fc741f68fe6714"
}
