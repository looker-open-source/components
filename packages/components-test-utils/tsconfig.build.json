{
  "extends": "../../tsconfig.build.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./lib"
  },
  "include": ["../../types", "src/**/*"]
}
