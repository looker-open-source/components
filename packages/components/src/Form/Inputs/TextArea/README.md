# TextArea

`TextArea` component offers a multi-line text editor. If you are building a form `FieldTextArea` may be a better fit.

```tsx
function Basic(props: TextAreaProps) {
  return <TextArea {...props} placeholder="Placholder text" />;
}
```

## DefaultValue

```tsx
function DefaultValue() {
  return <TextArea defaultValue="A value" />;
}
```

## Allowing TextArea to be resized

`true` and `vertical` will allow the `TextArea` height to be resized. Alternatively, `none` and `false` will not allow resizing.

```tsx
function Resize() {
  return (
    <>
      <TextArea resize placeholder="resize vertically" />
      <TextArea resize="none" placeholder="no resize" />
      <TextArea resize={false} placeholder="no resize" />
      <TextArea resize="vertical" placeholder="only resize vertically" />
    </>
  );
}
```

## Disabled

```tsx
function Disabled() {
  return <TextArea disabled />;
}
```

## Error Message

```tsx
function Error() {
  return <TextArea validationType="error" />;
}
```
