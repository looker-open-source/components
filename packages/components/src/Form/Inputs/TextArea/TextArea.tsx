/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import pick from 'lodash/pick'
import React, { FC } from 'react'
import styled from 'styled-components'
import {
  BorderProps,
  LayoutProps,
  SpaceProps,
  TypographyProps,
  CompatibleHTMLProps,
  layout,
  space,
  typography,
} from '@looker/design-tokens'
import { ValidationType } from '../../ValidationMessage'
import {
  inputTextHover,
  inputTextFocus,
  inputTextDisabled,
  inputTextValidation,
} from '../InputText'
import { Icon } from '../../../Icon'
import { inputPropKeys } from '../InputProps'

type TextAreaResize = 'vertical' | 'none' | boolean

export interface TextAreaProps
  extends BorderProps,
    Omit<LayoutProps, 'size'>,
    SpaceProps,
    TypographyProps,
    CompatibleHTMLProps<HTMLTextAreaElement> {
  /**
   * @default: vertical
   */
  resize?: TextAreaResize
  validationType?: ValidationType
}

const TextAreaLayout: FC<TextAreaProps> = ({
  className,
  validationType,
  ...props
}) => {
  const textareaProps = pick(props, inputPropKeys)

  return (
    <div className={className}>
      <textarea
        aria-invalid={validationType === 'error' ? 'true' : undefined}
        {...textareaProps}
      />
      {validationType && (
        <Icon name="CircleInfo" color="palette.red500" size={20} />
      )}
    </div>
  )
}

const TextAreaResize = (resize?: TextAreaResize) =>
  resize === false ? 'none' : resize === true ? 'vertical' : resize

export const TextArea = styled(TextAreaLayout)`
  height: fit-content;
  position: relative;
  width: 100%;

  ${Icon} {
    position: absolute;
    top: 0.5rem;
    right: 0.5rem;
  }

  textarea {
    ${layout}
    ${space}
    ${typography}

    resize: ${(props) => TextAreaResize(props.resize)};
    border: solid 1px;
    width: 100%;

    border-color: ${(props) => props.theme.colors.palette.charcoal200};
    border-radius: ${(props) => props.theme.radii.medium};
    font-size: ${(props) => props.theme.fontSizes.small};
    min-height: 6.25rem;
    padding: ${({ theme }) => `${theme.space.xsmall} ${theme.space.small}`};
    padding-right: ${(props) => props.theme.space.xlarge};

    &:hover {
      ${inputTextHover}
    }
    &:focus,
    :focus-within {
      ${inputTextFocus}
    }

    ${(props) => (props.disabled ? inputTextDisabled : '')}

    ${inputTextValidation}
  }
`

TextArea.defaultProps = {
  height: '6.25rem',
  resize: 'vertical',
  width: '100%',
}

TextArea.displayName = 'TextArea'
