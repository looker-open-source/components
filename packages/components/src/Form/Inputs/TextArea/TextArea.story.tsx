/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { Story } from '@storybook/react/types-6-0'
import { TextArea, TextAreaProps } from './TextArea'

export default {
  component: TextArea,
  title: 'TextArea',
}

const Template: Story<TextAreaProps> = (args) => <TextArea {...args} />

export const Basic = Template.bind({})

export const Placeholder = Template.bind({})
Placeholder.args = {
  placeholder: 'Placeholder',
}

export const Value = Template.bind({})
Value.args = {
  defaultValue: 'A value',
}

export const Disabled = Template.bind({})
Disabled.args = {
  ...Value.args,
  disabled: true,
}

export const Resize = Template.bind({})
Resize.args = {
  resize: true,
}

export const Error = Template.bind({})
Error.args = {
  validationType: 'error',
}
