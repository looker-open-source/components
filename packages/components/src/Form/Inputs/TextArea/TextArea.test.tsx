/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { renderWithTheme } from '@looker/components-test-utils'
import { screen } from '@testing-library/react'
import { TextArea } from './TextArea'

describe('TextArea', () => {
  test('with placeholder', () => {
    renderWithTheme(
      <TextArea id="TextAreaID" placeholder="this is a placeholder" />
    )
    expect(
      screen.getByPlaceholderText('this is a placeholder')
    ).toBeInTheDocument()
  })

  test('should accept disabled', () => {
    renderWithTheme(<TextArea disabled id="TextAreaID" />)
    expect(screen.getByRole('textbox')).toBeDisabled()
  })

  test('with an error validation', () => {
    renderWithTheme(<TextArea id="TextAreaID" validationType="error" />)
    expect(screen.getByRole('textbox')).toBeInvalid()
  })

  test('TextArea resizes with prop resize = true', () => {
    renderWithTheme(<TextArea id="TextAreaID" resize />)
    expect(screen.getByRole('textbox')).toHaveStyle('resize: vertical')
  })

  test('TextArea resizes with prop resize = false', () => {
    renderWithTheme(<TextArea id="TextAreaID" resize={false} />)
    expect(screen.getByRole('textbox')).toHaveStyle('resize: none')
  })

  test('TextArea resizes with prop resize = none', () => {
    renderWithTheme(<TextArea id="TextAreaID" resize="none" />)
    expect(screen.getByRole('textbox')).toHaveStyle('resize: none')
  })

  test('TextArea resizes with prop resize = vertical', () => {
    renderWithTheme(<TextArea id="TextAreaID" resize="vertical" />)
    expect(screen.getByRole('textbox')).toHaveStyle('resize: vertical')
  })
})
