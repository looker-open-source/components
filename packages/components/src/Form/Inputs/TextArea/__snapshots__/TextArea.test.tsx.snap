// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea default 1`] = `
.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: relative;
  width: 100%;
}

.c0 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c0 textarea {
  min-height: 6.25rem;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
  padding: 0.5rem 0.75rem;
  padding-right: 2rem;
  resize: vertical;
  width: 100%;
}

.c0 textarea:hover {
  border-color: #C1C6CC;
}

.c0 textarea:focus,
.c0 textarea:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    id="TextAreaID"
  />
</div>
`;

exports[`TextArea resizes with prop resize = false 1`] = `
.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: relative;
  width: 100%;
}

.c0 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c0 textarea {
  min-height: 6.25rem;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
  padding: 0.5rem 0.75rem;
  padding-right: 2rem;
  resize: none;
  width: 100%;
}

.c0 textarea:hover {
  border-color: #C1C6CC;
}

.c0 textarea:focus,
.c0 textarea:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    id="TextAreaID"
  />
</div>
`;

exports[`TextArea resizes with prop resize = none 1`] = `
.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: relative;
  width: 100%;
}

.c0 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c0 textarea {
  min-height: 6.25rem;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
  padding: 0.5rem 0.75rem;
  padding-right: 2rem;
  resize: none;
  width: 100%;
}

.c0 textarea:hover {
  border-color: #C1C6CC;
}

.c0 textarea:focus,
.c0 textarea:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    id="TextAreaID"
  />
</div>
`;

exports[`TextArea resizes with prop resize = true 1`] = `
.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: relative;
  width: 100%;
}

.c0 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c0 textarea {
  min-height: 6.25rem;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
  padding: 0.5rem 0.75rem;
  padding-right: 2rem;
  resize: vertical;
  width: 100%;
}

.c0 textarea:hover {
  border-color: #C1C6CC;
}

.c0 textarea:focus,
.c0 textarea:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    id="TextAreaID"
  />
</div>
`;

exports[`TextArea resizes with prop resize = vertical 1`] = `
.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: relative;
  width: 100%;
}

.c0 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c0 textarea {
  min-height: 6.25rem;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
  padding: 0.5rem 0.75rem;
  padding-right: 2rem;
  resize: vertical;
  width: 100%;
}

.c0 textarea:hover {
  border-color: #C1C6CC;
}

.c0 textarea:focus,
.c0 textarea:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    id="TextAreaID"
  />
</div>
`;

exports[`TextArea should accept disabled 1`] = `
.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: relative;
  width: 100%;
}

.c0 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c0 textarea {
  min-height: 6.25rem;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
  padding: 0.5rem 0.75rem;
  padding-right: 2rem;
  resize: vertical;
  width: 100%;
  background: #F5F6F7;
  color: #939BA5;
}

.c0 textarea:hover {
  border-color: #C1C6CC;
}

.c0 textarea:focus,
.c0 textarea:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c0 textarea:hover {
  border-color: #DEE1E5;
}

<div
  className="c0"
>
  <textarea
    disabled={true}
    id="TextAreaID"
  />
</div>
`;

exports[`TextArea with an error validation 1`] = `
.c2 {
  width: 20px;
  height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  color: #CC1F36;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 svg {
  height: 100%;
  width: 100%;
}

.c3 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c4 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: relative;
  width: 100%;
}

.c0 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c0 textarea {
  min-height: 6.25rem;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
  padding: 0.5rem 0.75rem;
  padding-right: 2rem;
  resize: vertical;
  width: 100%;
  border-color: #ED3B53;
}

.c0 textarea:hover {
  border-color: #C1C6CC;
}

.c0 textarea:focus,
.c0 textarea:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c0 textarea:hover {
  border-color: #CC1F36;
}

.c0 textarea:focus,
.c0 textarea:focus-within {
  border-color: #CC1F36;
  box-shadow: 0 0 0 2px #FFE5E9;
}

<div
  className="c0"
>
  <textarea
    aria-invalid="true"
    id="TextAreaID"
  />
  <div
    aria-hidden={true}
    className="c1 c2"
  >
    <svg
      fill="currentColor"
      height="100%"
      viewBox="0 0 24 24"
      width="100%"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`TextArea with placeholder 1`] = `
.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: relative;
  width: 100%;
}

.c0 .c1 {
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
}

.c0 textarea {
  min-height: 6.25rem;
  background: #FFFFFF;
  border: 1px solid #DEE1E5;
  border-radius: 0.25rem;
  color: #343C42;
  font-size: 0.875rem;
  padding: 0.5rem 0.75rem;
  padding-right: 2rem;
  resize: vertical;
  width: 100%;
}

.c0 textarea:hover {
  border-color: #C1C6CC;
}

.c0 textarea:focus,
.c0 textarea:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    id="TextAreaID"
    placeholder="this is a placeholder"
  />
</div>
`;
